contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,262 @@
+    function initialize() external initializer {
+        __CarbonVortex_init();
+    }
+
+    // solhint-disable func-name-mixedcase
+
+    /**
+     * @dev initializes the contract and its parents
+     */
+    function __CarbonVortex_init() internal onlyInitializing {
+        __Upgradeable_init();
+        __ReentrancyGuard_init();
+
+        __CarbonVortex_init_unchained();
+    }
+
+    /**
+     * @dev performs contract-specific initialization
+     */
+    function __CarbonVortex_init_unchained() internal onlyInitializing {
+        setRewards(100_000);
+    }
+
+    /**
+     * @dev authorize the contract to receive the native token
+     */
+    receive() external payable {}
+
+    /**
+     * @inheritdoc Upgradeable
+     */
+    function version() public pure override(IVersioned, Upgradeable) returns (uint16) {
+        return 2;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function setRewards(uint256 newRewardsPercentagePPM) public onlyAdmin validFee(uint32(newRewardsPercentagePPM)) {
+        uint256 prevRewardsPercentagePPM = _rewardsPercentagePPM;
+
+        // return if the rewards percentage PPM is the same
+        if (prevRewardsPercentagePPM == newRewardsPercentagePPM) {
+            return;
+        }
+
+        _rewardsPercentagePPM = newRewardsPercentagePPM;
+
+        emit RewardsUpdated({ prevRewards: prevRewardsPercentagePPM, newRewards: newRewardsPercentagePPM });
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function rewards() external view returns (uint256) {
+        return _rewardsPercentagePPM;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function totalBurnt() external view returns (uint256) {
+        return _totalBurnt;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function availableFees(Token token) external view returns (uint256) {
+        return _carbonController.accumulatedFees(token) + token.balanceOf(address(this));
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function execute(Token[] calldata tokens) external nonReentrant validateTokens(tokens) {
+        uint256 len = tokens.length;
+
+        // allocate balances array for the tokens
+        uint256[] memory balances = new uint256[](len);
+        // allocate array for the reward amounts for caller
+        uint256[] memory rewardAmounts = new uint256[](len);
+
+        // withdraw fees, load balances and reward amounts
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            // withdraw token fees
+            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
+            // get token balance
+            balances[i] = tokens[i].balanceOf(address(this));
+            // get reward amount for token
+            rewardAmounts[i] = MathEx.mulDivF(balances[i], _rewardsPercentagePPM, PPM_RESOLUTION);
+        }
+
+        // convert tokens to BNT
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
```

Comment:
Why do we need to split that into two loops?
Moreover, why do we need to build the `balances` array when only a single balance is relevant at each iteration, and the array itself isn't used outside the loop?

Reply:
Consider if it was one loop, in which we get the balance, user rewards and swap
the case when a user sends [TKN1, TKN2, BNT] : 
1. we get the token balance of TKN1 and allocate 2% user rewards for TKN1, swap 98% of TKN1 for BNT
2. we get the token balance of TKN2 and allocate 2% user rewards for TKN2, swap 98% of TKN2 for BNT
3. we get the token balance of BNT - which has increased based on the last two swaps significantly, and allocate 2% user rewards in BNT
You can see how the user can inflate his rewards by reordering the tokens he sends.
By using two loops, we make sure the user doesn't receive more rewards than he should.
We cache the balances so that we don't call `balanceOf` again in the second loop.

Reply:
Are you saying that the call to `_bancorNetwork.tradeBySourceAmount`  impacts the outcome of the next call to `tokens[i].balanceOf(address(this))`?

Reply:
Of course, since we trade the current token for BNT - so, the BNT balance of the contract increases.
Also, the current token balance decreases.

Reply:
But you are only using the balance of that token in the context of one specific iteration.
So how could `tradeBySourceAmount(bnt, tokens[i])` impact `tokens[i + 1].balanceOf(address(this))`?

Reply:
It does impact the next token if that token is BNT

Reply:
It is never BNT as far as I can see, because of the `if (token == _bnt) continue;` line a little bit further down the code.
But perhaps I'm missing some context here, so this is for you to consider (a single for loop, without the need to maintain an array of all the balances).

Reply:
The user rewards for BNT will be calculated before that check - since even if we don't swap, user still needs to receive his BNT rewards - where the entire issue comes from.